File: src/main/java/net/obesephoenix/koalakraftcrystals/commands/KKTabCompleter.java
package net.obesephoenix.koalakraftcrystals.commands;

import net.obesephoenix.koalakraftcrystals.util.KKTabCompletionUtil;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.util.StringUtil;

import java.util.ArrayList;
import java.util.List;

public class KKTabCompleter implements TabCompleter {
    @Override
    public List<String> onTabComplete(CommandSender commandSender, Command command, String label, String[] args) {
        List<String> options = new ArrayList<>();
        boolean found = false;

        if (label.equalsIgnoreCase("kkc")) {
            if (args.length != 0) {
                for (KKCommand cmd : KKCommandHandler.getCommands()) {
                    if (args[0].equalsIgnoreCase(cmd.getName())) {
                        options = cmd.tabComplete(commandSender, args);
                        found = true;
                    }
                }
                if (!found) StringUtil.copyPartialMatches(args[0], KKTabCompletionUtil.getCommands(), options);
            }
        }

        return options;
    }
}

File: src/main/kotlin/net/obesephoenix/koalakraftcrystals/commands/KKTabCompleter.kt
package net.obesephoenix.koalakraftcrystals.commands

import org.bukkit.command.TabCompleter
import net.obesephoenix.koalakraftcrystals.commands.KKCommand
import net.obesephoenix.koalakraftcrystals.commands.KKCommandHandler
import net.obesephoenix.koalakraftcrystals.util.KKTabCompletionUtil
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.util.StringUtil
import java.util.ArrayList

class KKTabCompleter : TabCompleter {

    override fun onTabComplete(
        commandSender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>
    ): List<String> {
        var options: MutableList<String> = ArrayList()
        var found = false
        if (label.equals("kkc", ignoreCase = true)) {
            if (args.isNotEmpty()) {
                for (cmd in KKCommandHandler.getCommands()) {
                    if (args[0].equals(cmd.name, ignoreCase = true)) {
                        options = cmd.tabComplete(commandSender, *args) as MutableList<String>
                        found = true
                    }
                }
                if (!found) StringUtil.copyPartialMatches(args[0], KKTabCompletionUtil.commands, options)
            }
        }
        return options
    }
}